### PR TITLE
fix(memh5): ensure contained numpy arrays have a native dtype.byteorder

### PR DESCRIPTION
**Pull request by anjakefala**
Needed because of https://github.com/mpi4py/mpi4py/issues/177.
Shift-Bcasted arrays with little or big endian byteorders will result in
an exception.

Is it ever possible that the endian set does not match the architecture? What do we do in that case?

If we use an upgraded version of mpi4py, this PR is not needed.

### CODE REVIEW

**Review by jrs65 — CHANGES_REQUESTED** [review]
The implementations of the conversion functions looks great.

However, I think all the conversions are being applied in the wrong directions. They seem to only be in the `*_to_hdf5` routines, and not in the `_from_hdf5` routines. The `_to_hdf5` routines won't be problematic as h5py will map everything to the LE types, but probably shouldn't have any conversion applied at all in case you do want to deliberately massage things.

The same is also true of `deep_group_copy`. That function has a bunch of branches such that it's a bit hard to follow, but I think it's only being applied when the destination is an `h5py.Group`, when it should actually be only applied if the conversion is `h5py.Group -> memh5.Group`.

So I think steps to do:
- [x] Remove conversions from `to_hdf5` routines.
- [ ] Add conversion to `from_hdf5` routine.
- [ ] Change `deep_group_copy` to make conversions happen only for `h5py -> memh5` copies.

**Comment by anjakefala** [comment]
Attempted to address comments!

File: caput/memh5.py [hunk 1]
```diff
@@ -2239,6 +2239,13 @@ def deep_group_copy(
             else:
                 data = entry[selection]
 
+            # only apply if conversion is h5py.Group -> memh5.Group
+            # reading from h5py can result in arrays with explicit endian set
+            # which mpi4py cannot handle when Bcasting memh5.Group
+            # needed until fixed: https://github.com/mpi4py/mpi4py/issues/177
+            if isinstance(g1, h5py.group) and isinstance(g2, MemGroup):
```
Comment (jrs65) [inline caput/memh5.py:2246]: There's obviously a typo here (lower case g), which is presumably why the tests fail.

However, to prepare for a zarr world I wonder if we should change the first clause to `not instance(g1, MemGroup)`.

Also, I'm not sure what zarr does in terms of typing of arrays, so maybe we don't need to do it.

File: caput/memh5.py [hunk 2]
```diff
@@ -2588,6 +2598,9 @@ def _copy_from_file(h5group, memgroup, selections=None):
 
                     cdata = comm.bcast(cdata, root=0)
 
+                    # only needed until fixed: https://github.com/mpi4py/mpi4py/issues/177
+                    cdata = ensure_native_byteorder(pdata)
```
Comment (jrs65) [inline caput/memh5.py:2602]: This should be `cdata` as the argument, and it should also probably be before the `bcast` call. Obviously it's `Bcast` that's been causing us trouble, but just in case things change, or we change `bcast -> Bcast` at somepoint.

**Review by jrs65 — CHANGES_REQUESTED** [review]

File: caput/memh5.py (hunk 1 above)
Comment (jrs65) [reply caput/memh5.py:2246]: Looks like `zarr` gives an array with native order, so we don't need to do anything at all. In which case your version (with the fixed typo) is all good.

File: caput/memh5.py (hunk 2 above)
Comment (anjakefala) [reply caput/memh5.py:2602]: Good catch on both counts! My eyes are getting tired.

**Comment by anjakefala** [comment]
```
>       return self._axis
E       AttributeError: 'MPIArray' object has no attribute '_axis'
```

*grieves*

I think we should merge in the MPIArray work first, instead of coding in the workarounds.

File: caput/memh5.py [hunk 3]
```diff
@@ -2569,6 +2576,9 @@ def _copy_from_file(h5group, memgroup, selections=None):
                         h5group, key, axis=distributed_axis, comm=comm, sel=selection
                     )
 
+                    # only needed until fixed: https://github.com/mpi4py/mpi4py/issues/177
+                    pdata = ensure_native_byteorder(pdata)
```
Comment (jrs65) [inline caput/memh5.py:2580]: I'm pretty sure this call is the one that's causing the test failures, but it actually isn't needed as `.from_hdf5` will guarantee that pdata has a explicit native endianness. Maybe true removing it, and see if things work in that case?

**Review by jrs65 — CHANGES_REQUESTED** [review]

File: caput/memh5.py [hunk 4]
```diff
@@ -2586,6 +2593,9 @@ def _copy_from_file(h5group, memgroup, selections=None):
                         if convert_dataset_strings:
                             cdata = ensure_unicode(cdata)
 
+                    # only needed until fixed: https://github.com/mpi4py/mpi4py/issues/177
+                    cdata = ensure_native_byteorder(cdata)
```
Comment (anjakefala) [inline caput/memh5.py:2597]: Only rank 0 will have `cdata`.

**Review by jrs65 — APPROVED** [review]
